### PR TITLE
[SLO Routing] Add Latency Predictor sidecars and EPP tools

### LINKED ISSUE #1323: [Feat] SLO-Aware Routing with Latency Prediction
SLO Aware Routing is a strategy to satisfy individual request TTFT and TPOT latency SLOs leveraging latency prediction to optimize pod selection.

We propose several experimental changes to the request flow, including a new scheduling profile, training and prediction sidecars, and live request tracking for each pod in datastore.

## EPP Deployment Requirements
1. Include the latency prediction flag in EPP runtime flags
2. Add latency and training sidecars as containers to deployment yaml
3. Include 2 scheduling profiles, one for just latency prediction, and another for combined prediction and routing

Example scheduling profiles:
```
apiVersion: inference.networking.x-k8s.io/v1alpha1
    kind: EndpointPickerConfig
    plugins:
    - type: queue-scorer
    - type: kv-cache-utilization-scorer
    - type: prefix-cache-scorer
    - type: slo-request-tracker
    - type: slo-scorer
    - type: slo-aware-profile-handler
    - type: weighted-random-picker
    schedulingProfiles:
    - name: default
      plugins:
      - pluginRef: slo-request-tracker
      - pluginRef: queue-scorer
      - pluginRef: kv-cache-utilization-scorer
      - pluginRef: prefix-cache-scorer
    - name: slo
      plugins:
      - pluginRef: prefix-cache-scorer
        weight: 0
      - pluginRef: slo-request-tracker
      - pluginRef: slo-scorer
      - pluginRef: weighted-random-picker
```

New Plugins:

1. **slo-aware-profile-handler**
Chooses the scheduling profile depending on the PredictionBasedRouting boolean header value.

2. **slo-request-tracker**
This handles the request tracking logic for latency prediction, adding requests to the list of running requests on pods as they are scheduled, and interacts with sidecards, making prediction requests and sending the required training data after requests complete.

3. **slo-scorer**
Performs the scoring for Prediction Based Routing, based on the latency prediction of a request for each pod, chooses the best pod to schedule the request to.

4. **weighted-random-picker**
A picker to choose randomly between many different pods based on weights assigned to them from the scorers.

## Request Flow

The request flow for will be as follows:
1. Request received by gateway and may have TTFT and TPOT SLOs and PredictionBasedRouting in the header.
2. `slo-aware-profile-handler` checks PredictionBasedRouting header: if True, proceed to use an SLO-aware scheduling profile instead of the default:
a. For each potential pod, run latency prediction and store it in memory along the request path.
b. Identify "valid" pods, defined as the pods predicted to serve the request within its SLO and within the SLO of all running requests, or have no running requests (checking the list of running requests on each pod).

**Prediction Based Routing:** (done in `slo-scorer` plugin):
- **if len(valid_pods) > 0**: Return a weighted random draw favoring pods (done by `weighted-random-picker`) with the lowest OR highest positive headroom based on a "scheduling strategy" environment variable:
-Lowest: Assign to pods that have just enough resources to meet SLO, maintaining pods with high headroom for large critical requests
-Highest: Assign to pods that have substantial resources to meet SLO, so as to evenly distribute load.
(Both options, perhaps a very small chance of choosing an invalid pod, for exploration for training purposes)
- **else if len(valid_pods) == 0**:
-**if request is critical**: Return a weighted random draw favoring pods with the lowest negative headroom (least “overwhelmed” pods among those not meeting SLO).
-**else if request is NOT critical**: shed request [Saturation Detection]

3. Once a pod is decided, store the request with predicted ttft/tpot in datastore under that pods running requests
4. Forward request to the selected pod endpoint, `slo-request-tracker` adds it to the datastore running request queue for that pod.
5. Continuously add the history of actual latencies and predicted latencies to the running requests on the pod in the datastore.
6. Once a request completes, `slo-request-tracker` removes it from the queue.

## Diagram

Image

### Headroom

We define Headroom not as the difference between a specific requests SLO and its predicted latency, but rather the difference between the lowest SLO of all the running requests on the pod.

Example:
Say we want to serve a request with 80ms TPOT SLO.

pod-1 has a running request with 50ms TPOT SLO (likely currently being served at <50ms TPOT), predictor says that scheduling this new request would have a predicted latency of 60ms TPOT, scorer would then consider this pod to have -10ms of headroom, not 20ms

If there are no other pods with positive headroom and this request is non-critical then we shed, as it is predicted to violate (i.e. no pod has positive headroom for it), but if it's critical we schedule it to the least negative headroom pod.

### Pods in Datastore
We add predictions and SLO to a new priority queue of running requests on each pod, stored as:
- Request ID
- SLO
- Predicted TTFT
- Predicted TPOT

### Criticality
Criticality will be handled by the layer above scheduling, allowing the scheduler to focus on efficient bin-packing. The saturation detector will be responsible for shedding non-critical requests if SLOs cannot be met.
### Prefix Score Reuse
The new SLO-aware profile can reuse the existing prefix score logic.
### No SLO Provided
If the latency prediction flag is enabled in EPP, we require all requests to provide an SLO, error if otherwise. We decided on this behavior since a mixture of SLO and non-SLO requests will serve the non-SLO requests better, which is counter intuitive since by specifying a latency target you are expecting the algorithm to try and meet it. It is best to set this expectation as a requirement.

## v0 Implementation

For our initial implementation, the only feature we don't intend to incorporate is the Saturation Detection inside the SLO profile. This is because upcoming flow control changes will change how criticality behaves and we want to maintain compatibility with that new system.

### PR DESCRIPTION
**Pull request by BenjaminBraunDev**
This is stage 1 of 3 for adding in the latency prediction functionality to EPP.

Adds a new directory for the latency prediction and training sidecars for upcoming experimental latency prediction and SLO based routing scheduling profiles. Full details in issue below.

This does not change nor add any functionality as is, only introduces the sidecar module directory to be used to build the training and prediction images. These will later be deployed as containers alongside the EPP image.

Fixes #1323 

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: latencypredictor-v1/README.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+## Configuration
+
+Before running the script, update the configuration variables in `build-deploy.sh`:
+
+```bash
+# Edit these values in the script
+PROJECT_ID="your-gcp-project-id"
+REGION="your-gcp-region"
+REPOSITORY="your-artifact-registry-repo"
+TRAINING_IMAGE="latencypredictor-v2-training-server"
```
Comment (kfswain) [inline latencypredictor-v1/README.md]: This name doesn't match the directory name, which is more of a nit

But it made me realize that we should probably just name this directory `latencypredictor` no need to have a separate versioning system compared to the rest of the system.

File: latencypredictor-v1/manifests/dual-server-deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    metadata:
+      labels:
+        app: training-server
+        component: training
+    spec:
+      nodeSelector:
+        cloud.google.com/gke-nodepool: "pool-2"
+      containers:
+      - name: training-server
+        image: us-docker.pkg.dev/kaushikmitra-gke-dev/kaushikmitra-docker-repo/latencypredictor-v3-training-server:latest
```
Comment (kfswain) [inline latencypredictor-v1/manifests/dual-server-deployment.yaml]: need to convert this image to something more standard and user agnostic

File: latencypredictor-v1/manifests/dual-server-deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    metadata:
+      labels:
+        app: prediction-server
+        component: prediction
+    spec:
+      nodeSelector:
+        cloud.google.com/gke-nodepool: "pool-2"
+      containers:
+      - name: prediction-server
+        image: us-docker.pkg.dev/kaushikmitra-gke-dev/kaushikmitra-docker-repo/latencypredictor-v3-prediction-server:latest
```
Comment (kfswain) [inline latencypredictor-v1/manifests/dual-server-deployment.yaml]: same here

File: latencypredictor-v1/manifests/test-dual-server-deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        component: test
+    spec:
+      # Use the same node pool as your services
+      nodeSelector:
+        cloud.google.com/gke-nodepool: "pool-2"
+      restartPolicy: Never
+      containers:
+      - name: test-runner
+        # Use your test image here
+        image: us-docker.pkg.dev/kaushikmitra-gke-dev/kaushikmitra-docker-repo/latencypredictor-v3-test:latest
```
Comment (kfswain) [inline latencypredictor-v1/manifests/test-dual-server-deployment.yaml]: same here

File: latencypredictor-v1/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+scikit-learn
+numpy
+pandas
+joblib
+river
+pydantic
+requests
+xgboost
+aiohttp
+lightgbm
```
Comment (kfswain) [inline latencypredictor-v1/requirements.txt]: mega-nit: newline at end of file. linters oft complain about this

File: pkg/epp/latencypredictorasync/latencypredictor_async.go [hunk 6]
```diff
@@ -0,0 +1,1312 @@
...
+		config = ConfigFromEnv()
+	}
+	p := &Predictor{
+		config:     config,
+		httpClient: &http.Client{Timeout: config.HTTPTimeout},
+		logger:     logger.WithName("latency-predictor-client"),
+		rng:        rand.New(rand.NewSource(time.Now().UnixNano())),
+		done:       make(chan struct{}),
+	}
+	p.wg.Add(1)
```
Comment (kfswain) [inline pkg/epp/latencypredictorasync/latencypredictor_async.go:62]: We should have this wg increment in the same func the Done()/decrement is called to prevent a confusing read (this happens in `backgroundLoop()`). Is there a reason we seperated them here?

File: pkg/epp/latencypredictorasync/latencypredictor_async.go [hunk 7]
```diff
@@ -0,0 +1,1312 @@
...
+			cancel()
+		case <-p.done:
+			return
+		}
+	}
+}
+
+// GetServerStatus fetches the current status from a prediction server
+func (p *Predictor) GetServerStatus(ctx context.Context) (*ServerStatusResponse, error) {
+	if err := p.refreshServerStatus(ctx); err != nil {
```
Comment (kfswain) [inline pkg/epp/latencypredictorasync/latencypredictor_async.go]: Why do we call this here instead of relying on the background loop? Also, is this function used anywhere?

File: pkg/epp/latencypredictorasync/latencypredictor_async.go [hunk 8]
```diff
@@ -0,0 +1,1312 @@
...
+	var status ServerStatusResponse
+	if err := json.NewDecoder(resp.Body).Decode(&status); err != nil {
+		return fmt.Errorf("failed to decode server status response: %w", err)
+	}
+
+	p.metricsMu.Lock()
+	p.serverStatus = &status
+	p.metricsMu.Unlock()
+
+	p.logger.V(1).Info("Retrieved server status",
```
Comment (kfswain) [inline pkg/epp/latencypredictorasync/latencypredictor_async.go]: reccomend using the semantic levels for logging defined here: https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/pkg/epp/util/logging/logging_const.go#L20 and changing this to a more verbose log level

File: sidecars/latencypredictorasync/latencypredictor_async.go [hunk 9]
```diff
@@ -0,0 +1,1312 @@
+// Package latencypredictorasync provides a Go client for the Python-based
+// latency prediction service with asynchronous batching and cached metrics.
+package latencypredictorasync
```
Comment (kfswain) [inline sidecars/latencypredictorasync/latencypredictor_async.go:17]: should probably breaking this logic out to separate files within this package to facilitate readability.

File: latencypredictor-v1/requirements.txt (hunk 5 above)
Comment (BenjaminBraunDev) [reply latencypredictor-v1/requirements.txt]: done

File: latencypredictor-v1/README.md (hunk 1 above)
Comment (BenjaminBraunDev) [reply latencypredictor-v1/README.md]: Removed the v2 naming

File: latencypredictor-v1/manifests/dual-server-deployment.yaml (hunk 2 above)
Comment (BenjaminBraunDev) [reply latencypredictor-v1/manifests/dual-server-deployment.yaml]: replaced with a placeholder image string for now

File: latencypredictor-v1/manifests/dual-server-deployment.yaml (hunk 3 above)
Comment (BenjaminBraunDev) [reply latencypredictor-v1/manifests/dual-server-deployment.yaml]: same as above

File: latencypredictor-v1/manifests/test-dual-server-deployment.yaml (hunk 4 above)
Comment (BenjaminBraunDev) [reply latencypredictor-v1/manifests/test-dual-server-deployment.yaml]: same as above

File: pkg/epp/latencypredictorasync/latencypredictor_async.go (hunk 6 above)
Comment (BenjaminBraunDev) [reply pkg/epp/latencypredictorasync/latencypredictor_async.go:62]: moved into the background loop

File: pkg/epp/latencypredictorasync/latencypredictor_async.go (hunk 7 above)
Comment (BenjaminBraunDev) [reply pkg/epp/latencypredictorasync/latencypredictor_async.go]: I believe this is entirely unused, I asked Kaushik and he confirmed. Removed this and related tests.

File: pkg/epp/latencypredictorasync/latencypredictor_async.go (hunk 8 above)
Comment (BenjaminBraunDev) [reply pkg/epp/latencypredictorasync/latencypredictor_async.go]: Done, using DEBUG and TRACE for levels 1 and 2 respectively

File: sidecars/latencypredictorasync/latencypredictor_async.go (hunk 9 above)
Comment (BenjaminBraunDev) [reply sidecars/latencypredictorasync/latencypredictor_async.go:17]: Done, broke code out into metrics, prediction, training, and types

File: latencypredictor/__pycache__/test_dual_server_client.cpython-312-pytest-8.4.1.pyc
Comment (kfswain) [inline latencypredictor/__pycache__/test_dual_server_client.cpython-312-pytest-8.4.1.pyc:1]: Are these binary files required to be checked in?

File: latencypredictor/build-deploy.sh [hunk 10]
```diff
@@ -0,0 +1,226 @@
+#!/bin/bash
+# Build and deploy script for both servers
+
+set -e
+
+# Configuration
+PROJECT_ID="kaushikmitra-gke-dev" # REPLACE with your GCP project ID
```
Comment (kfswain) [inline latencypredictor/build-deploy.sh]: pls remove dev project dependencies

File: sidecars/latencypredictorasync/latencypredictor_async.go
Comment (kfswain) [inline sidecars/latencypredictorasync/latencypredictor_async.go:1]: I think this plugin implementation should not be at the `pkg/epp/.` level. Can we consider creating a new directory for plugin implementations to live? It can be something simple since this is moderately temporary.

File: pkg/epp/latencypredictorasync/latencypredictor_async.go [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+	defer p.metricsMu.RUnlock()
+	if p.cachedMetrics == nil {
+		return nil, false
+	}
+	return p.cachedMetrics, true
+}
+
+// IsXGBoostReady returns true if native XGBoost models are loaded and ready.
+func (p *Predictor) IsXGBoostReady() bool {
+	p.xgboostMu.RLock()
```
Comment (kfswain) [inline pkg/epp/latencypredictorasync/latencypredictor_async.go]: AFAICT, this is the only place this lock is used. Is it needed? Esp since it looks like these values could be modified out from underneath this lock

**Review by kfswain — COMMENTED** [review]
Made another pass! 

I think most importantly, we need to remove the binary files unless there is a good reason to check them in. 

Otherwise left some smaller comments, and it's looking good.

File: latencypredictor/__pycache__/test_dual_server_client.cpython-312-pytest-8.4.1.pyc
Comment (BenjaminBraunDev) [reply latencypredictor/__pycache__/test_dual_server_client.cpython-312-pytest-8.4.1.pyc:1]: Ah, I remove the v1 from the directory name and forgot to change the .gitignore while doing that. No, these should not be checked in, let me fix that.

File: sidecars/latencypredictorasync/latencypredictor_async.go
Comment (BenjaminBraunDev) [reply sidecars/latencypredictorasync/latencypredictor_async.go:1]: This is not the actual plugin implementation yet, we have all our plugin logic in the multi plugin folder (which is planned for the next PR). `latencypredictor_async` is more or less the suite of methods the plugin uses to interface with the python sidecar. I will create a `sidecars` directory at the root level and put it there.

File: pkg/epp/latencypredictorasync/latencypredictor_async.go (hunk 11 above)
Comment (BenjaminBraunDev) [reply pkg/epp/latencypredictorasync/latencypredictor_async.go]: Yeah this looks to be unnecessary, removing.

File: latencypredictor/build-deploy.sh (hunk 10 above)
Comment (BenjaminBraunDev) [reply latencypredictor/build-deploy.sh]: done

File: latencypredictor/__pycache__/test_dual_server_client.cpython-312-pytest-8.4.1.pyc
Comment (BenjaminBraunDev) [reply latencypredictor/__pycache__/test_dual_server_client.cpython-312-pytest-8.4.1.pyc:1]: done
